test cases/kconfig/3 load_config files/dir/meson.build

```diff
@@ -0,0 +1,13 @@
+
+k = import('unstable-kconfig')
+
+conf = k.load(files('config'))
```

Comment:
I have not used kconfig all that much, but shouldn't we load the _output_ file? Specifically something that will get generated in the build dir rather than reading stuff from the source dir? Loading stuff from source dir means that you can have only one configuration per source dir, not one per build dir as is the Meson standard.

Reply:
> I have not used kconfig all that much, but shouldn't we load the output file? Specifically something that will get generated in the build dir rather than reading stuff from the source dir? Loading stuff from source dir means that you can have only one configuration per source dir, not one per build dir as is the Meson standard.

Indeed, since Meson only supports out-of-tree build, it is somewhat unfortunate that the default for `kconfig` output is `.config` in the toplevel source directory. However, the output file can be overridden using an environment variable, usually specified like

    $ make menuconfig KCONFIG_CONFIG=foo.config

I expect that users of this module will make the location of the configuration file a Meson option too, so that the same file can be specified both as `KCONFIG_CONFIG` and when invoking Meson. I will add a note to the documentation.

Of course, having to specify the directory twice is ugly; unfortunately on one hand the build directory is not known when you invoke the kconfig frontend, and on the other hand the kconfig frontend must be invoked before Meson. One possible fix is to provide some kind of script that wraps `meson setup` and invokes kconfig before Meson (removing the need to ship a `Makefile`), not unlike those Meson users that are providing a simple-minded `configure` script. That script can then pass a file in the build directory and pass it to both kconfig and Meson.

That said, the placement of the configuration also depends on the usecase. For example, QEMU only uses kconfig as a common format to store pre-defined `configuration_data` for each binary; it does not employ any front-end, instead all processing of the configuration happens in `meson.build`. In that case, the configuration files are fixed and committed, and they definitely come from the source directory.

Reply:
Since the module is unstable, we can change it later. We can merge this and change the behaviour as people test it on real world code.

Reply:
Maybe something like:

```meson
k.generate_and_load(init_command: ['kconfig', 'source_file', 'output_file'])
```

That runs the given command if the output `.config` file does not exist in the build dir and otherwise just loads it?

Reply:
It seems sensible but I would first try and see how people use this. The config frontends usually have their own make-based build system, and invoking make from Meson would be a bit weird...

Reply:
> Maybe something like:
> k.generate_and_load(init_command: ['kconfig', 'source_file', 'output_file'])

If you were referring to: `genkconfig .config -> config.h`, then yes, I wrote a `custom_target` that does just that.

If you were referring to generating the default `.config` automatically with `olddefconfig` then no, it's very easy but apparently considered bad practice in the (interactive) Kconfig world to automagically build the default configuration. This must be an explicit choice from the user as seen in the Linux kernel example.